packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+  /// configuration. The configuration will be used by [routerDelegate] and
+  /// eventually rebuilds the [Router] widget.
+  ///
+  /// Since this delegate is the primary consumer of the [routeInformationProvider],
+  /// it must not be null if [routeInformationProvider] is not null.
+  final RouteInformationParser<T> routeInformationParser;
+
+  /// The router delegate for the router.
+  ///
+  /// This delegate consumes the configuration from [routeInformationParser] and
+  /// builds a navigating widget for the [Router].
+  ///
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// Retrieve the router of exact type.
+  static Router<R> of<R>(BuildContext context) {
+    final _RouterScope scope = context.dependOnInheritedWidgetOfExactType<_RouterScope>();
+    assert(scope != null);
+    return scope.routerState.widget as Router<R>;
+  }
+
+  /// Forces the [Router] to to run the [callback] with the assumption whether
+  /// it is a navigating event.
+  ///
+  /// Normally, calls to this method is not necessary because the [Router] will
+  /// detect whether the callback is a navigating event automatically as soon
+  /// as it is run. Only use this method when you want to force the [Router]
+  /// to recognize or ignore the [callback] as navigation event.
+  ///
+  /// When [router] detect a navigating event, it will report that event to the
+  /// [routeInformationProvider]. This is useful in the web application because
+  /// the web platform relies on these reporting to support browser backward and
+  /// forward buttons.
+  ///
+  /// if the [isNavigationEvent] is true, the [Router] will treat the [callback]
+  /// as a navigating event regardless and report the event to the
+  /// [routeInformationProvider].
+  ///
+  /// if the [isNavigationEvent] is false, the [Router] will treat the [callback]
+  /// as a non-navigating event regardless and will not report to the
+  /// [routeInformationProvider].
+  ///
+  /// if nothing is passed into [isNavigatingEvent], the [Router] will
+  /// decide whether the [callback] is a navigating event by itself.
+  static void run(
+    BuildContext context,
+    VoidCallback callback, {
+    bool isNavigatingEvent,
+  }) {
+    final _RouterScope scope = context.getElementForInheritedWidgetOfExactType<_RouterScope>().widget as _RouterScope;
+
+    if (isNavigatingEvent == true)
+      scope.routerState._reportingStatus = _RouteInformationReportingStatus.must;
+    if (isNavigatingEvent == false)
+      scope.routerState._reportingStatus = _RouteInformationReportingStatus.ignore;
+    scope.routerState.setState(callback);
```

Comment:
Should routerState have a dedicated API where you pass in the reporting status and the callback at once? Seems odd that these informaiton are passed in through different paths.

Also, does _reportingStatus need to be reset after the callback?